jax/_src/scipy/special.py

```diff
@@ -2092,14 +2106,114 @@ def expi(x: ArrayLike) -> Array:
   (x,) = primals
   (x_dot,) = tangents
   return expi(x), jnp.exp(x) / x * x_dot
 
+@custom_derivatives.custom_jvp
+@jit
+def sici(x: ArrayLike) -> tuple[Array, Array]:
+  r"""Sine and cosine integrals.
+
+  JAX implementation of :obj:`scipy.special.sici`.
+
+  .. math::
+
+    \mathrm{Si}(x) = \int_0^x \frac{\sin t}{t} \, dt
+
+  .. math::
+
+    \mathrm{Ci}(x) = \gamma + \ln(x) + \int_0^x \frac{\cos t - 1}{t} \, dt
+
+  where :math:`\gamma` is the Euler–Mascheroni constant.
+
+  Args:
+    x: array-like, real-valued input.
+
+  Returns:
+    A tuple of two arrays, each with the same shape as `x`:
+      - The first array contains the sine integral values `Si(x)`.
+      - The second array contains the cosine integral values `Ci(x)`.
+
+  See also:
+    - :func:`jax.numpy.sinc`
+  """
+
+  x, = promote_args_inexact("sici", x)
+
+  if dtypes.issubdtype(x.dtype, np.complexfloating):
+    raise ValueError(
+      f"Argument `x` to sici must be real-valued. Got dtype {x.dtype}."
+    )
+
+  def si_series(x):
+    # Values come from Cephes Implementation used by Scipy https://github.com/jeremybarnes/cephes/blob/60f27df395b8322c2da22c83751a2366b82d50d1/misc/sici.c
+    SN = np.array([-8.39167827910303881427E-11,
+      4.62591714427012837309E-8,
+      -9.75759303843632795789E-6,
+      9.76945438170435310816E-4,
+      -4.13470316229406538752E-2,
+      1.00000000000000000302E0], dtype=x.dtype)
+    SD = np.array([ 2.03269266195951942049E-12,
+      1.27997891179943299903E-9,
+      4.41827842801218905784E-7,
+      9.96412122043875552487E-5,
+      1.42085239326149893930E-2,
+      9.99999999999999996984E-1], dtype=x.dtype)
+    t = x * x
+    return (x * jnp.polyval(SN, t)) / jnp.polyval(SD, t)
+
+  def ci_series(x):
+    # Values come from Cephes Implementation used by Scipy https://github.com/jeremybarnes/cephes/blob/60f27df395b8322c2da22c83751a2366b82d50d1/misc/sici.c
+    CN = np.array([ 2.02524002389102268789E-11,
+      -1.35249504915790756375E-8,
+      3.59325051419993077021E-6,
+      -4.74007206873407909465E-4,
+      2.89159652607555242092E-2,
+      -1.00000000000000000080E0], dtype=x.dtype)
+    CD = np.array([ 4.07746040061880559506E-12,
+      3.06780997581887812692E-9,
+      1.23210355685883423679E-6,
+      3.17442024775032769882E-4,
+      5.10028056236446052392E-2,
+      4.00000000000000000080E0], dtype=x.dtype)
+    t = x * x
+    return np.euler_gamma + jnp.log(x) + t * jnp.polyval(CN, t) / jnp.polyval(CD, t)
+
+
+  sign = jnp.sign(x)
+  xa = jnp.abs(x)
+
+  si = jnp.piecewise(
+    xa,
+    [xa == 0, jnp.isinf(xa)],
+    [0.0, np.pi/2, si_series]
+  )
+
+  ci = jnp.piecewise(
+    xa,
+    [x == 0, isposinf(x), isneginf(x)],
+    [-np.inf, 0.0, np.nan, ci_series]
+  )
+
+  si = sign * si
+
+  return si, ci
+
+@sici.defjvp
+@jit
+def sici_jvp(primals, tangents):
+  (p,), (t,) = primals, tangents
+  primal_out = sici(p)
+
+  sin_term = sinc(p / np.pi)
+  cos_term = jnp.where(p != 0, jnp.cos(p)/p, 0.0)
```

Comment:
Change this to
```python
cos_term = jnp.cos(p) / p
```
The expression diverges at `p = 0`, so we should not make it return `0.0` in this case.

Reply:
Since it's undefined at 0, won't this cause an issue?

Reply:
Also will this cause any kind of inconsistency since the left limit is not the same as the right limit?

Reply:
> Since it diverges at 0, won't this cause an issue? At the very least it looks like a divide by zero.

Evaluating a diverging function at its point of divergence could cause issues, yes. That doesn't mean we should replace the diverging value with an incorrect output.

Reply:
Wait isn't that what the sinc function does? It too should be undefined at zero but instead either substitutes 1 or does an approximation.

Reply:
Huh interesting, I didn't realize numpy scalars implement IEEE 754 floating-point while python doesn't

np.cos(0.0)/0.0 => Infinity
1.0/0.0 => error

I guess then infinity is the return value here and that's fine? Perhaps this is worth mentioning in the function definition somehow? I.e. the derivative is not defined at zero or returns infinity at zero? I feels odd to not return the power series for sinc, but to fall back to floating point defaults for cosc.